### ISSUE TITLE
Using collectionView.rx_itemsWithCellFactory on Swift 3 will crash with error "fatal error: DataSource not set"

### ISSUE THREAD

**Issue by siuying** [ReactiveX/RxSwift#825]
*Short description*:

On Swift 3 (Xcode 8 beta 4), when I try to bind observable to `collectionView.rx_itemsWithCellFactory`, the app will crash with error "fatal error: DataSource not set" .

*Self contained code example that reproduces the issue*:

```swift
class DataCell: UICollectionViewCell {
    @IBOutlet var value: UILabel?
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        
        value = UILabel(frame: contentView.bounds)
        value?.autoresizingMask = [.flexibleHeight, .flexibleWidth]
        contentView.addSubview(value!)
    }
}

class ViewController: UIViewController {
    var collectionView: UICollectionView!
    private let disposeBag = DisposeBag()

    override func loadView() {
        super.loadView()

        let layout = UICollectionViewFlowLayout()
        layout.estimatedItemSize = CGSize(width: 160, height: 240)
        self.collectionView = UICollectionView(frame: self.view.bounds, collectionViewLayout: layout)
        self.collectionView!.collectionViewLayout = layout
        self.collectionView!.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        self.collectionView!.backgroundColor = UIColor.white
        self.view.addSubview(self.collectionView!)
        
        self.collectionView!.register(DataCell.classForCoder(), forCellWithReuseIdentifier: "Cell")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        let items = [
            "Hello",
            "World",
            "Foo",
            "Bar",
        ]
        
        Observable
            .just(items)
            .bindTo(collectionView.rx_itemsWithCellFactory) { collectionView, row, title in
                let indexPath = IndexPath(item: row, section: 0)
                guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath) as? DataCell else {
                    fatalError("missing cell")
                }
                cell.value?.text = title
                return cell
            }
            .addDisposableTo(disposeBag)
    }
}
```

Sample project: https://github.com/siuying/RxSwift_825

  *Xcode version*:

  Xcode 8 beta 4

*Expected outcome*:

The app should not crash.

*What actually happens*:

It crash with error "fatal error: DataSource not set"

On Swift 2.2 the above code will not crash. Similar setup table view will not crash too.

**Comment by siuying** [comment]
Yes the fix solve the issue. Thank you for the great work!

**Comment by entropious** [comment]
Still an issue for pod 'RxCocoa', '~> 2.0'

**Comment by kzaher** [comment]
Hi @entropious ,

we've released 3.0.0 that solves this issue.

**Comment by entropious** [comment]
I know, but we are still using swift 2.3 and currently can't use swift 3.0 pod. What about 2.x branch, you don't support it any more?

**Comment by kzaher** [comment]
Hi @entropious ,

if you want to use Swift 2.3, fix for this is in https://github.com/ReactiveX/RxSwift/tree/rxswift-2.0 branch. So if you target it directly it should work for you.

**Comment by entropious** [comment]
Thank you for your answers. I tried 
    pod 'RxSwift', :git => 'https://github.com/ReactiveX/RxSwift', :branch => 'rxswift-2.0'
    pod 'RxCocoa', :git => 'https://github.com/ReactiveX/RxSwift', :branch => 'rxswift-2.0'
but getting same crash.
In master CollectionViewDataSourceNotSet returns 0 for numberOfItemsInSection (https://github.com/ReactiveX/RxSwift/blob/master/RxCocoa/iOS/Proxies/RxCollectionViewDataSourceProxy.swift), in rxswift-2.0 CollectionViewDataSourceNotSet crashes on assert (https://github.com/ReactiveX/RxSwift/blob/rxswift-2.0/RxCocoa/iOS/Proxies/RxCollectionViewDataSourceProxy.swift)

**Comment by kzaher** [comment]
Hi @entropious ,

I've pushed fix in rxswift-2.0 branch. I think this might solve your issue.

**Comment by entropious** [comment]
Thank you, it works now!
